### PR TITLE
CosmosChainProcessor query loop

### PR DESCRIPTION
**Pull request by agouin**
Adds main query loop for `CosmosChainProcessor`

`Run` function will:
1. Initially, before starting loop, query for latest height, retrying until is successful. 

I went back and forth on whether this should return an error which would stop the `CosmosChainProcessor` or not, since once in the loop, we want it to continue to retry in the case of a node outage incident. But maybe initially it should panic?

2. Determine initial block to query from provided `initialBlockHistory`
3. Start the query loop

Each iteration of the query loop:
1. Query latest height
2. If not yet in sync, check if `latestHeight` - `latestQueriedBlock` is within `inSyncThreshold`. If determined to now be in sync, update concurrently accessed `inSync` property to true.
3. Query `BlockResults` from each block from `latestQueriedBlock + 1` to `latestHeight`. This loop defined as `BlockQuery` below.
4. If this is the first query loop iteration where `inSync == true`, trigger the associated `PathProcessors` to check if both chains are `InSync()`, and start processing the backlog of IBC messages if so.
5. Check how long this cycle of the query loop took. If less than `minQueryLoopDuration`, wait until `minQueryLoopDuration` has passed from the start of the cycle before continuing. TODO in a future PR, make `minQueryLoopDuration` dynamic based on block production time.

For each `BlockQuery` iteration:
1. Iterate through transactions in block.
2. Use event parser code from previous PR to pull out parsed IBC messages from successful transactions.
3. TODO handle IBC messages in future PR

### CODE REVIEW

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 1]
```diff
@@ -11,22 +16,45 @@ import (
...
+	}, nil
 }
 
+const (
+	latestHeightQueryTimeout    = 5 * time.Second
+	latestHeightQueryRetryDelay = 1 * time.Second
+	latestHeightQueryRetries    = 5
+
+	defaultMinQueryLoopDuration = 1 * time.Second
+	inSyncThreshold             = 2 //blocks
```
Comment (DavidNix) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: nit:

`inSyncNumBlocksThreshold`?

I think it's clearer to put any unit or denomination in the variable/constant name itself. When you encounter the constant elsewhere, you don't have to refer back to the declaration hoping someone commented.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 2]
```diff
@@ -11,22 +16,45 @@ import (
...
+)
+
 // InSync indicates whether queries are in sync with latest height of the chain.
 // The PathProcessors use this as a signal for determining if the backlog of messaged is ready to be processed and relayed.
 func (ccp *CosmosChainProcessor) InSync() bool {
-	return false
+	ccp.inSyncLock.RLock()
+	defer ccp.inSyncLock.RUnlock()
+	return ccp.inSync
```
Comment (DavidNix) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: For threadsafe bools, I typically use [the atomic package](https://pkg.go.dev/sync/atomic) and an int32 where 0 is false and 1 is true. I feel it's less error prone than mutexes expecially RW mutexes. 

You can then abstract the logic into methods like.

```go
func (ccp *CosmosChainProcessor) InSync() bool {
   return atomic.LoadInt32(&ccp.inSync) == 1
}

func (ccp *CosmosChainProcessor) SetInSync(inSync bool) {
  if inSync {
    atomic.StoreInt32(&ccp.inSync, 1)
  // ... and the inverse, storing 0 for false
}
```

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 3]
```diff
@@ -40,9 +68,147 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
...
+
+	ccp.log.Debug("Queried latest height",
+		zap.String("chain_id", chainID),
+		zap.Int64("latest_height", persistence.latestHeight),
+	)
+
+	// Only need read lock here
+	ccp.inSyncLock.RLock()
+	inSync := ccp.inSync
+	ccp.inSyncLock.RUnlock()
```
Comment (DavidNix) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: Why not use public `InSync` method here? Or call it directly on line 162.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 4]
```diff
@@ -40,9 +68,147 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
...
+
+	// used at the end of the cycle to send signal to path processors to start processing if both chains are in sync and no new messages came in this cycle
+	firstTimeInSync := false
+
+	if !inSync {
+		if (persistence.latestHeight - persistence.latestQueriedBlock) < inSyncThreshold {
+			// take write lock to update to true (only happens once)
+			ccp.inSyncLock.Lock()
+			ccp.inSync = true
+			ccp.inSyncLock.Unlock()
```
Comment (DavidNix) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: A setter may be safer here `ccp.setInSync(bool)`. Inside that method can use the sync primitive so callers need not worry about it.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 5]
```diff
@@ -40,9 +68,147 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
...
+
+	// Only need read lock here
+	ccp.inSyncLock.RLock()
+	inSync := ccp.inSync
+	ccp.inSyncLock.RUnlock()
+
+	// used at the end of the cycle to send signal to path processors to start processing if both chains are in sync and no new messages came in this cycle
+	firstTimeInSync := false
+
+	if !inSync {
```
Comment (DavidNix) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: Because this is a local variable, we may risk a race condition where inSync changes by the time we get here and the local var is stale.

Not sure if that race matters though.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 6]
```diff
@@ -40,9 +68,147 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
...
-	panic("not yet implemented")
+	// this will be used for persistence across query cycle loop executions
+	queryCyclePersistence := QueryCyclePersistence{
+		minQueryLoopDuration: defaultMinQueryLoopDuration,
+	}
+
+	// Infinite retry to get initial latest height
+	for {
+		latestHeight, err := ccp.latestHeightWithRetry(ctx)
+		if err == nil {
```
Comment (DavidNix) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: Log error?

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 7]
```diff
@@ -40,9 +68,147 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
...
+	ccp.log.Info("Entering main query loop", zap.String("chain_id", ccp.ChainProvider.ChainId()))
+
+	for {
+		if ctx.Err() != nil {
+			return nil
+		}
+		if err := ccp.queryCycle(ctx, &queryCyclePersistence); err != nil {
+			return err
+		}
+	}
```
Comment (DavidNix) [inline relayer/chains/cosmos/cosmos_chain_processor.go:140]: This seems like we could use a for select loop with a ticker and could eliminate `doneWithCycle` sleep. 

https://gobyexample.com/tickers

Comment (DavidNix) [reply relayer/chains/cosmos/cosmos_chain_processor.go:140]: A ticker will also make the rate limit more apparent. When reading line 130 it's not obvious there is a delay via time.sleep.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 8]
```diff
@@ -40,9 +68,147 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
...
+		zap.Int64("latest_height", persistence.latestHeight),
+	)
+
+	// Only need read lock here
+	ccp.inSyncLock.RLock()
+	inSync := ccp.inSync
+	ccp.inSyncLock.RUnlock()
+
+	// used at the end of the cycle to send signal to path processors to start processing if both chains are in sync and no new messages came in this cycle
+	firstTimeInSync := false
```
Comment (DavidNix) [inline relayer/chains/cosmos/cosmos_chain_processor.go:162]: I don't see where this is ever flipped to true?

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 5 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: This loop is the only writer, so won't be concurrent writes. Only reads otherwise.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 6 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: `latestHeightWithRetry` will log the error(s)

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 8 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go:162]: line 169

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 6 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: Actually it won't log the final error looks like, will update.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 8 above)
Comment (DavidNix) [reply relayer/chains/cosmos/cosmos_chain_processor.go:162]: Ah sorry, missed it. My browser search failed me.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 5 above)
Comment (DavidNix) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: If it's the only writer, why protect `ccp.inSync` from concurrent access?

**Comment by DavidNix** [comment]
I bet there's a control loop / single writer pattern that could fit here as well. But it would take me a while to refactor toward that. Happy to take a shot at it if it seems worthwhile.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 5 above)
Comment (DavidNix) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: Oh, I see now. Reads may happen concurrently because `InSync()` is an exported method.

Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go:162]: Readers could collide with a write.

**Comment by agouin** [comment]
> I bet there's a control loop / single writer pattern that could fit here as well. But it would take me a while to refactor toward that. Happy to take a shot at it if it seems worthwhile.

As a heads up, there are more concurrently accessed variables coming in future PRs, `InSync()` is the first. Happy to consider other patterns for these, but I do want the `PathProcessors` to be able to grab the latest cached values immediately (or as soon as possible after a write in the case that a write is happening when the value is requested) from the `ChainProcessors` without having to wait for the next iteration of the `ChainProcessor` loops to handle the response.

https://github.com/cosmos/relayer/blob/71320fc6677eecab222f2ac5fd372a4b98517aec/relayer/chains/cosmos/cosmos_chain_processor.go#L59

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 1 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: done

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 3 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: done

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 4 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: done

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 6 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: done

**Comment by DavidNix** [comment]
You're right - concurrent reads changes the game. 

> but I do want the PathProcessors to be able to grab the latest cached values immediately 

One alternate patten is PathProcessors are *sent* the latest values such as via a channel. That way they are ignorant of the the ChainProcessor. 

Depending on how long it takes to process something, I'm not sure that will be viable though. In other words, how up to date do we have to be?

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 7 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go:140]: I was planning on changing the rate limit dynamically from the block queries by measuring the block production time, in a future PR. With a ticker, it looks like the way to handle this is to stop the existing ticker and recreate a new one. I'm open to either though.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 2 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: done

**Comment by agouin** [comment]
> One alternate patten is PathProcessors are sent the latest values such as via a channel

I like this idea, this could work well. But don't we still come back to channels vs mutexes within the PathProcessor, so really just a shift in responsibility for handling the concurrently accessed vars?

> how up to date do we have to be?

I would like for it to be as up to date as possible to avoid unnecessary relayer wallet spends.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 7 above)
Comment (DavidNix) [reply relayer/chains/cosmos/cosmos_chain_processor.go:140]: Ticker has a [reset method](https://pkg.go.dev/time#Ticker.Reset), so you don't need to stop/recreate.

**Comment by DavidNix** [comment]
[Use atomic int32 access instead of RWMutex for bool](https://github.com/cosmos/relayer/pull/765/commits/93226cf7ff1b97782a3134309e174e5b4d2b6f4a)

If there's more state beyond a single bool to cache in the future, this may be the wrong pattern. And you can blame me. 😄

File: relayer/chains/cosmos/sdk_helper.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+	client, err := newClient(rpcAddress)
+	if err != nil {
+		return nil, err
+	}
+	codec := makeCodec([]module.AppModuleBasic{})
+	return &cosmosClient.Context{
+		Client:            client,
+		ChainID:           chainID,
+		Input:             os.Stdin,
+		Output:            os.Stdout,
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/sdk_helper.go]: These should be threaded in through a cobra command if at all possible.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 10]
```diff
@@ -40,9 +71,151 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
...
+			"Failed to query latest height",
+			zap.String("chain_id", ccp.ChainProvider.ChainId()),
+			zap.Uint("attempt", n+1),
+			zap.Uint("max_attempts", latestHeightQueryRetries),
+			zap.Error(err),
+		)
+	}))
+}
+
+type QueryCyclePersistence struct {
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: Looks like this type can and should be unexported.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 11]
```diff
@@ -11,22 +16,48 @@ import (
 type CosmosChainProcessor struct {
 	log *zap.Logger
 
+	ChainProvider *cosmos.CosmosProvider
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: Does this need to be exported?

**Review by mark-rushakoff — COMMENTED** [review]
It's hard to follow what's happening here when I don't have a good mental model of how the path processors interact with the chain processors.

The bidirectionality implied here is concerning, because it introduces a large surface area for things that can go wrong.

Assuming we had a relayer that was working with three chains and two paths like this:

```
┌───────────────┐                   ┌───────────────┐                     ┌───────────────┐
│     gaia      │◀────gaia-osmo────▶│    osmosis    │◀─────osmo-juno─────▶│     juno      │
└───────────────┘                   └───────────────┘                     └───────────────┘
```

Then I would expect the object relationships, and the communication direction, to look like this:

```
┌─────────────────┐                ┌─────────────────┐                   ┌─────────────────┐
│                 │                │                 │                   │                 │
│                 │                │                 │                   │                 │
│                 │                │                 │                   │                 │
│ ChainProcessor: │                │ ChainProcessor: │                   │ ChainProcessor: │
│      gaia       │                │     osmosis     │                   │      juno       │
│                 │                │                 │                   │                 │
│                 │                │                 │                   │                 │
│                 │                │                 │                   │                 │
│                 │                │                 │                   │                 │
└─────────────────┘                └─────────────────┘                   └─────────────────┘
         │                                  │                                     │         
         │                                  │                                     │         
         │                                  │                                     │         
         │                                  │                                     │         
         │        ┌─────────────────┐       │         ┌─────────────────┐         │         
         │        │                 │       │         │                 │         │         
         │        │                 │       │         │                 │         │         
         │        │                 │       │         │                 │         │         
         │        │ PathProcessor:  │       │         │ PathProcessor:  │         │         
         └───────▶│    gaia-osmo    │◀──────┴────────▶│    osmo-juno    │◀────────┘         
                  │                 │                 │                 │                   
                  │                 │                 │                 │                   
                  │                 │                 │                 │                   
                  │                 │                 │                 │                   
                  └─────────────────┘                 └─────────────────┘                   
```

The chain processors are aware of the path processors, because individual events/transactions observed on a chain need to be sent to the appropriate path processor. 

But does the path processor really need to query the chain processor? Can't the chain processor say, "here's a transaction for block 1... here's another transaction for block 1... okay, I'm fully synced on block 1"; and that is the signal to relay everything needed for block 1?

If that is oversimplified for what must happen, then a diagram of some sort would be really helpful for me to understand the intent here.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 11 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: not now, but the `PathProcessor` will need it in the future.

**Comment by agouin** [comment]
Here's a happy path packet flow with this architecture so we can go through the interactions:

For this example, user is sending IBC transfer from `PathEnd1` to `PathEnd2`

1. User broadcasts `MsgTransfer` transaction to `PathEnd1` chain
2. Relayer observes this `MsgTransfer` (`CosmosChainProcessor` picks it up out of `ibcMessagesFromTransaction`), and passes the information needed to construct the `MsgRecvPacket` for the `PathEnd2` chain to the relevant `PathProcessor`.
3. `PathProcessor` `processLatestMessages` determines that, based on the latest state of the `MessageCache`, there is a `MsgTransfer` for this packet's sequence number on the `PathEnd1` chain, but there is not yet a `MsgRecvPacket` for this packet's sequence number on the `PathEnd2` chain. There is also not a packet-flow-complete message (`MsgAcknowledgement`, `MsgTimeout`, or `MsgTimeoutOnClose`) for this packet's sequence number on the `PathEnd1` chain, which would indicate it was already completely handled.
4. `PathProcessor`, based on information in 3, seeing that both `PathEnd1` and `PathEnd2` are in sync (querying the latest blocks), and also seeing that the packet timeout has not expired, determines that it needs to send `MsgRecvPacket` to `PathEnd2` chain.

Note: up to this point, the `PathProcessor` has not needed any additional information from the `ChainProcessors` other than the incoming messages and whether the chains are in sync.

5. To construct the `MsgRecvPacket` IBC message for the `PathEnd2` chain, it needs to query the `PathEnd1` chain for the packet commitment to prove that the funds have been sent by the user. This is a query that does not come from cached data, it has to hit the RPC node for the `PathEnd1` chain. So no concerns about concurrent access of cached data here.
6. To send the `MsgRecvPacket` to the `PathEnd2` chain, it also needs to include a prepended `MsgUpdateClient` message to update the IBC client so that the clients stay in sync between the 2 chains (See _Constructing MsgUpdateClient_ below).
7. Broadcast transaction to `PathEnd2` chain with the constructed messages `[MsgUpdateClient, MsgRecvPacket]`.

Upon successful transaction, packet is successfully relayed.

8. `ChainProcessor` for `PathEnd2` chain will observe the messages in a committed block, and pass the information needed to construct a `MsgAcknowledgement` for the `PathEnd1` chain to the `PathProcessor`.
9. `PathProcessor` `processLatestMessages` determines that, based on the latest state of the `MessageCache`, there was a `MsgTransfer` for this packet's sequence number on the `PathEnd1` chain, there was a `MsgRecvPacket` for this packet's sequence number on the `PathEnd2` chain, but there is not yet a `MsgAcknowledgement` message for this packet's sequence number on the `PathEnd1` chain.
10. `PathProcessor`, based on information in 9, determines that it needs to send `MsgAcknowledgement` to `PathEnd1` chain.
11. To construct the `MsgAcknowledgement` IBC message for the `PathEnd1` chain, it needs to query the `PathEnd2` chain for the packet acknowledgement commitment to prove that the packet has been successfully relayed to the `PathEnd2` chain. This is a query that does not come from cached data, it has to hit the RPC node for the `PathEnd2` chain. So no concerns about concurrent access of cached data here.
12. To send the `MsgAcknowledgement` to the PathEnd1 chain, it also needs to include a prepended `MsgUpdateClient` message to update the IBC client so that the clients stay in sync between the 2 chains (See _Constructing MsgUpdateClient_ below, but flip `PathEnd1` and `PathEnd2`).
13. Broadcast transaction to `PathEnd1` chain with constructed messages `[MsgUpdateClient, MsgAcknowledgement]`.

Upon successful transaction, packet is successfully acked

14. `ChainProcessor` for `PathEnd1` chain will observe the messages in a committed block, and pass the information to the `PathProcessor`.
15. `PathProcessor` `processLatestMessages` determines that, based on the latest state of the `MessageCache`, packet flow is complete for the packet's sequence. All data for the specific packet sequence is wiped from the cache.

### Constructing MsgUpdateClient
To construct a `MsgUpdateClient` for the `PathEnd2` chain, we can use cached data that originated from the `ChainProcessors`:

1. Get the latest client trusted height from `PathEnd2` chain (the client used for this path). This could be a query to the RPC node for the `PathEnd2` chain, but it's unnecessary since we observe all IBC client messages in the `ChainProcessors`.
2. Get the validator set and signed block header from the latest block on the `PathEnd1` chain. This could be a query to the RPC node for the `PathEnd1` chain, but it's unnecessary since we observe all blocks and can cache the necessary data from the latest block.
3. Get the validator set from the `PathEnd1` chain's block at the `PathEnd2`'s client trusted height. This could be a query to the RPC node for the `PathEnd1` chain, but it's unnecessary since we observe all blocks and can cache the recent validator sets.
4. Construct `MsgUpdateClient` for the `PathEnd2` chain with the client ID, signing account, and a `Header` property constructed using the data retrieved in steps 1-3.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 10 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: done

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 11 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: changed this to be private. Since PathProcessors will need access to the relayer.ChainProvider, accessing that is now with a `Provider` interface method.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 12]
```diff
@@ -40,9 +60,147 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
...
+}
+
+func (ccp *CosmosChainProcessor) queryCycle(ctx context.Context, persistence *queryCyclePersistence) error {
+	chainID := ccp.chainProvider.ChainId()
+
+	var err error
+	persistence.latestHeight, err = ccp.latestHeightWithRetry(ctx)
+
+	// don't want to cause CosmosChainProcessor to quit here, can retry again next cycle.
+	if err != nil {
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/cosmos_chain_processor.go:148]: I'm not sure if this needs to special-case `context.Canceled` or `context.DeadlineExceeded`, but that's always easy to add later.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 13]
```diff
@@ -40,9 +60,147 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
...
+	// used at the end of the cycle to send signal to path processors to start processing if both chains are in sync and no new messages came in this cycle
+	firstTimeInSync := false
+
+	if !ccp.inSync {
+		if (persistence.latestHeight - persistence.latestQueriedBlock) < inSyncNumBlocksThreshold {
+			ccp.inSync = true
+			firstTimeInSync = true
+			ccp.log.Info("chain is in sync", zap.String("chain_id", chainID))
+		} else {
+			ccp.log.Info("chain is not yet in sync",
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: Log messages should begin with an uppercase letter.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 14]
```diff
@@ -2,36 +2,56 @@ package cosmos
...
 }
 
-func NewCosmosChainProcessor(log *zap.Logger, provider *cosmos.CosmosProvider, pathProcessors processor.PathProcessors) *CosmosChainProcessor {
+func NewCosmosChainProcessor(log *zap.Logger, provider *cosmos.CosmosProvider, rpcAddress string, pathProcessors processor.PathProcessors) (*CosmosChainProcessor, error) {
+	cc, err := getCosmosClient(rpcAddress, provider.ChainId())
+	if err != nil {
+		return nil, fmt.Errorf("error getting cosmos client: %w", err)
+	}
 	return &CosmosChainProcessor{
 		log:            log,
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: Consider `log: log.With(zap.String("chain_id", provider.ChainId()))` so that we don't need to repeat that on each log call.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 15]
```diff
@@ -40,9 +60,147 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
...
+			ccp.log.Info("chain is not yet in sync",
+				zap.String("chain_id", chainID),
+				zap.Int64("latest_queried_block", persistence.latestQueriedBlock),
+				zap.Int64("latest_height", persistence.latestHeight),
+			)
+		}
+	}
+
+	for i := persistence.latestQueriedBlock + 1; i <= persistence.latestHeight; i++ {
+		blockRes, err := ccp.cc.Client.BlockResults(ctx, &i)
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/cosmos_chain_processor.go:178]: I wonder if we can use `ccp.cc.Client.BlockSearch` to get several blocks in one API call. But that can happen separately, if it works.

File: relayer/processor/path_processor.go [hunk 16]
```diff
@@ -81,33 +84,42 @@ func (pp *PathProcessor) PathEnd2Messages(message string) SequenceCache {
...
+		pp.handleNewCacheDataForPathEnd(pp.pathEnd1, cacheData)
 	} else if pp.pathEnd2.info.ChainID == chainID {
 		// TODO make sure passes channel filter for pathEnd2 before calling this
-		pp.handleNewMessagesForPathEnd(pp.pathEnd2, messages)
+		pp.handleNewCacheDataForPathEnd(pp.pathEnd2, cacheData)
 	}
 }
 
-func (pp *PathProcessor) handleNewMessagesForPathEnd(
+func (pp *PathProcessor) handleNewCacheDataForPathEnd(
```
Comment (mark-rushakoff) [inline relayer/processor/path_processor.go]: I would probably this method in its two calls above, since it is just one channel send.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 17]
```diff
@@ -40,9 +60,147 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
...
-	panic("not yet implemented")
+	// this will be used for persistence across query cycle loop executions
+	persistence := queryCyclePersistence{
+		minQueryLoopDuration: defaultMinQueryLoopDuration,
+	}
+
+	// Infinite retry to get initial latest height
+	for {
+		latestHeight, err := ccp.latestHeightWithRetry(ctx)
+		if err != nil {
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/cosmos_chain_processor.go:102]: I think this block needs to check `errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded)` -- without that check, if the context is terminated during initial height retrieval, this method will loop forever.

File: relayer/processor/path_processor.go [hunk 18]
```diff
@@ -81,33 +84,42 @@ func (pp *PathProcessor) PathEnd2Messages(message string) SequenceCache {
...
+		pp.pathEnd2.chainProvider = chainProvider
 		return true
 	}
 	return false
 }
 
+// ProcessBacklogIfReady gives ChainProcessors a way to trigger the path processor process
+// as soon as they are in sync for the first time, even if they do not have new messages.
+func (pp *PathProcessor) ProcessBacklogIfReady() {
+	pp.retryProcess <- struct{}{}
```
Comment (mark-rushakoff) [inline relayer/processor/path_processor.go]: This will block if the `retryProcess` channel is already at capacity.

I think a slightly better approach is a select:

```go
select {
case pp.retryProcess <- struct{}{}:
  // All good.
default:
  // Log that the channel is saturated;
  // something is wrong if we are retrying this quickly.
  pp.log.Info("Failed to enqueue path processor retry")
}
```

Logging seems appropriate here, as it is likely an issue with the code if the channel is full.

**Review by mark-rushakoff — COMMENTED** [review]
I think this looks about complete -- `(*CosmosChainProcessor).Run` needs to handle context cancellation in the initial height loop, and it would still be good to thread in the cobra command's stdin and stdout in this PR rather than later.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 15 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go:178]: `BlockSearch` could definitely be helpful in the case of the initial flush if we wanted to avoid using the existing flush strategy. For typical operation after the initial flush, this will be scheduled so that `persistence.latestHeight - persistence.latestQueriedBlock` will be either `0` or `1`

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 17 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go:102]: good catch, done.

File: relayer/processor/path_processor.go (hunk 16 above)
Comment (agouin) [reply relayer/processor/path_processor.go]: done

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 14 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: This is very handy. Done

File: relayer/processor/path_processor.go (hunk 18 above)
Comment (agouin) [reply relayer/processor/path_processor.go]: done. Called this for the existing retry case also.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 13 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: done

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 12 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go:148]: will get caught by the select next time around.

File: relayer/chains/cosmos/sdk_helper.go (hunk 9 above)
Comment (agouin) [reply relayer/chains/cosmos/sdk_helper.go]: parameterized them through the `CosmosChainProcessor` instantiation.

**Review by mark-rushakoff — APPROVED** [review]
